osquery/tables/yara/yara.cpp

```diff
@@ -140,24 +152,41 @@ QueryData genYara(QueryContext& context) {
       YR_RULES* tmp_rules = nullptr;
       auto status = compileSingleFile(path, &tmp_rules);
       if (!status.ok()) {
-        VLOG(1) << "YARA compile error: " << status.toString();
+        LOG(WARNING) << "YARA compile error: " << status.toString();
         continue;
       }
       // Cache the compiled rules by setting the unique signature file path
       // as the lookup name. Additional signature file uses will skip the
       // compile step and be added as rule groups.
       rules[file] = tmp_rules;
     }
-    // Assemble an "ad-hoc" group using the signature file path as the name.
-    groups.insert(file);
+    // Add signature file path to the sign group set
+    sign_group.insert(std::make_pair(YC_FILE, file));
+  }
+
+  for (const auto& rule_string : sigrules) {
+    // Check if the yara rule string has not been compiled before
+    if (rules.count(rule_string) == 0) {
+      YR_RULES* tmp_rules = nullptr;
+      // Compile rule string and load it into rule pointer
+      auto status = compileFromString(rule_string, &tmp_rules);
+      if (!status.ok()) {
+        LOG(WARNING) << "YARA rule : " << rule_string << status.toString();
+        continue;
+      }
+
+      rules[rule_string] = tmp_rules;
```

Comment:
I feel like this caching could be dangerous, strings can be extremely large and people might run arbitrary queries via the distributed APIs. I suggest that we hash the rule string before adding it to the map here, add a known prefix, and that we clear these and free the memory after the scanning occurs.

Reply:
Thanks, @theopolis for the review. I updated it to hash the rule string before adding to the cache and remove it after the scan to reduce/avoid the possibility of collision in case of multiple queries raised by distributed APIs. This is similar to not adding the `sigrule` string to the cache. Please let me know your thoughts.